src/memory.rs

```diff
@@ -1,6 +1,59 @@
 use std::mem;
 use std::slice;
 
+/// A view into an externally owned byte slice (Go `[]byte`).
+/// Use this for the current call only. A view cannot be copied for safety reasons.
+/// If you need a copy, use [`to_owned`].
+///
+/// Go's nil value is fully supported, such that we can differentiate between nil and an empty slice.
+#[repr(C)]
+pub struct ByteSliceView {
+    /// True if and only if the byte slice is nil in Go. If this is true, the other fields must be ignored.
+    is_nil: bool,
+    ptr: *const u8,
+    len: usize,
+}
+
+impl ByteSliceView {
+    /// ByteSliceViews are only constructed in Go. This constructor is a way to mimic the behaviour
+    /// when testing FFI calls from Rust. It must not be used in production code.
+    #[cfg(test)]
+    pub fn new(source: &[u8]) -> Self {
+        Self {
+            is_nil: false,
+            ptr: source.as_ptr(),
+            len: source.len(),
+        }
+    }
+
+    /// ByteSliceViews are only constructed in Go. This constructor is a way to mimic the behaviour
+    /// when testing FFI calls from Rust. It must not be used in production code.
+    #[cfg(test)]
+    pub fn nil() -> Self {
+        Self {
+            is_nil: true,
+            ptr: std::ptr::null::<u8>(),
+            len: 0,
+        }
+    }
+
+    /// Provides a reference to the included data to be parsed or copied elsewhere
+    /// This is safe as long as the `ByteSliceView` is constructed correctly.
+    pub fn read(&self) -> Option<&[u8]> {
+        if self.is_nil {
+            None
```

Comment:
Cool, we can differentiate now. The wish for 12 months now

Reply:
Yeah, this can be done for Buffer as well. But one step at a time.